### PR TITLE
Stream json in from gzip files to reduce memory load

### PR DESCRIPTION
**Pull request by mweiden**
Fixes https://github.com/HumanCellAtlas/logs/issues/56

Before this change the whole s3 file submitted by firehose was downloaded to memory before uploading the data to elasticsearch. This PR makes the pipeline a stream, uploading 1000 long lines to elasticsearch before polling for more data.

Before
screen shot 2018-12-11 at 10 23 48 am

After
screen shot 2018-12-11 at 10 23 29 am

### CODE REVIEW

File: apps/firehose_to_es_processor/lib/gzip_stream_reader.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import io
+import zlib
+from enum import Enum
+from itertools import chain
+
+
+class GzipStreamReader(io.RawIOBase):
+
```
Comment (ttung) [inline apps/firehose_to_es_processor/lib/gzip_stream_reader.py]: Why not https://docs.python.org/3/library/gzip.html#gzip.GzipFile? This also presents a stream.

Comment (mweiden) [reply apps/firehose_to_es_processor/lib/gzip_stream_reader.py]: Ah, I missed this. Will simplify.

File: apps/firehose_to_es_processor/app.py [hunk 2]
```diff
@@ -58,20 +59,25 @@ def handler(event, context):
...
             notifier = AirbrakeNotifier()
             log_event_stream = notifier.notify_on_stream(log_event_stream)
 
-        log_events = list(log_event_stream)
         es_client = ESClient()
         es_client.create_cwl_day_index()
-        es_client.bulk_post(log_events)
+
+        total = 0
+        for batch_iter in group(1000, log_event_stream):
```
Comment (ttung) [inline apps/firehose_to_es_processor/app.py]: This whole construct is a bit strange to me.  `group` calls `itertools.islice`, which yields an iterator and forms a tuple out of it.  Then the caller turns it into a list.

I almost would rather that this group method be inlined.

Also, 1000 is a magic constant.

File: apps/firehose_to_es_processor/lib/json_object_stream.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import json
+
+
+class JsonObjectStream:
```
Comment (ttung) [inline apps/firehose_to_es_processor/lib/json_object_stream.py]: DANGER DANGER.  Would not want to build a json parser.  Does something like https://github.com/isagalaev/ijson satisfy your requirements?

Comment (mweiden) [reply apps/firehose_to_es_processor/lib/json_object_stream.py]: So if you look at this closely, all it does is determine boundaries of json objects. `json.loads` is still used to actually parse the json. Does this allay your fears?

**Comment by mweiden** [comment]
@ttung I've now confirmed that this system can batch 1e6 log lines per minute (~16.6k qps equiv).

File: apps/firehose_to_es_processor/lib/json_object_stream.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        slash_active = False
+        quotes_active = False
+        result = ''
+        while True:
+            c = self.reader.read(1)
+            if c == '':
+                return None
+            result += c
+            if c == '"':
+                quotes_active = quotes_active if slash_active else (False if quotes_active else True)
```
Comment (ttung) [inline apps/firehose_to_es_processor/lib/json_object_stream.py]:
```suggestion
                quotes_active = quotes_active if slash_active else (not quotes_active)
```

File: apps/firehose_to_es_processor/lib/json_object_stream.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    def next(self):
+        brackets = 0
+        slash_active = False
+        quotes_active = False
+        result = ''
+        while True:
+            c = self.reader.read(1)
+            if c == '':
+                return None
+            result += c
```
Comment (ttung) [inline apps/firehose_to_es_processor/lib/json_object_stream.py]: string concatenation in python is expensive.

File: apps/firehose_to_es_processor/lib/json_object_stream.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+            if c == '"':
+                quotes_active = quotes_active if slash_active else (False if quotes_active else True)
+            if c == '{':
+                brackets += 0 if quotes_active else 1
+            if c == '}':
+                brackets += 0 if quotes_active else -1
+                if brackets == 0:
+                    return json.loads(result)
+            slash_active = (False if slash_active else True) if c == "\\" else False
```
Comment (ttung) [inline apps/firehose_to_es_processor/lib/json_object_stream.py]:
```suggestion
            slash_active = (not slash_active) if c == "\\" else False
```

File: apps/firehose_to_es_processor/lib/json_object_stream.py (hunk 3 above)
Comment (ttung) [reply apps/firehose_to_es_processor/lib/json_object_stream.py]: ummm, not really?  :)

just not a fan of hand-written parsers without a buuuuunch of tests.

Comment (mweiden) [reply apps/firehose_to_es_processor/lib/json_object_stream.py]: I discussed this with Tony in person

**Pull request merged**